### ISSUE TITLE
SvgDocument.Open fails returning null sometimes when used by multiple threads

### ISSUE THREAD

**Issue by kimmccoy** [vvvv/SVG#59]
I am using Svg in a Asp.net Mvc app. Png images are generated via this library from SVG markup. It's a web app, so multiple threads hit the image generation code at the same time.
Sometimes, SvgDocument.Open will return a null even though the svg is valid. Other times it's fine. So not thread safe is my guess.

The problem is in SvgElementFactory.SetPropertyValue where more than one thread may attempt to set the same property value at the same time, attempting to add to _propertyDescriptors just after another thread has already added the dictionary key.

Easy fix is to add [MethodImpl(MethodImplOptions.Synchronized)] on the method.

SvgElementFactory.AvailableElements is also prone to this some problem.

Note: I don't have time to submit a nicely tested fix. My hack is to lock in my own app to avoid the issue. Hope this saves someone time. I will submit a patch if the lock becomes a performance issue for me.

**Comment by Threetwosevensixseven** [comment]
+1
Same problem here, in a MVC app.

My workaround is:

```
private static object _lock = new object();
...
SvgDocument doc = null;
lock (_lock)
{
     doc = SvgDocument.Open(file);
}
```

**Comment by Threetwosevensixseven** [comment]
I logged the first chance exceptions, and there's a whole bunch of stuff getting wadded up, culminating in the NullReferenceException:

```
System.NotSupportedException: Must call AddTerm AddSeparator in that order
   at ExCSS.TermList.AddTerm(Term term) in c:\Dev\vvvv\SVG\Source\External\ExCSS\Model\Values\TermList.cs:line 35
===========

System.NotSupportedException: Must call AddTerm AddSeparator in that order
   at ExCSS.TermList.AddTerm(Term term) in c:\Dev\vvvv\SVG\Source\External\ExCSS\Model\Values\TermList.cs:line 35
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: The CreateDocument method can only be used to parse root <svg> elements.
   at Svg.SvgElementFactory.CreateDocument[T](XmlReader reader) in c:\Dev\vvvv\SVG\Source\SvgElementFactory.cs:line 59
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
===========

System.NullReferenceException: Object reference not set to an instance of an object.
   at Svg.SvgDocument.Open[T](String path, Dictionary`2 entities) in c:\Dev\vvvv\SVG\Source\SvgDocument.cs:line 182
===========

System.NullReferenceException: Object reference not set to an instance of an object.
   at Svg.SvgDocument.Open[T](String path, Dictionary`2 entities) in c:\Dev\vvvv\SVG\Source\SvgDocument.cs:line 182
===========

System.NullReferenceException: Object reference not set to an instance of an object.
   at Svg.SvgDocument.Open[T](String path, Dictionary`2 entities) in c:\Dev\vvvv\SVG\Source\SvgDocument.cs:line 182
   at Svg.SvgDocument.Open(String path) in c:\Dev\vvvv\SVG\Source\SvgDocument.cs:line 146
   at OperationSmileyFrontEnd.Controllers.ImageController.RenderSVG(String Name, String Color, Int32 Height, Int32 Monochrome, String Mode) in c:\Users\Robin\Documents\Visual Studio 2010\Projects\TeamVoice\OperationSmileyFrontEnd\Controllers\ImageController.cs:line 36
   at lambda_method(Closure , ControllerBase , Object[] )
   at System.Web.Mvc.ActionMethodDispatcher.Execute(ControllerBase controller, Object[] parameters)
   at System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters)
   at System.Web.Mvc.ControllerActionInvoker.<>c__DisplayClass13.<InvokeActionMethodWithFilters>b__10()
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionMethodFilter(IActionFilter filter, ActionExecutingContext preContext, Func`1 continuation)
```

**Comment by JohnMcAvinue** [comment]
+1 on the same issue and suggested work around by @kimmccoy worked perfectly. Thanks!

**Comment by pulsatrixbv** [comment]
Any news regarding this? Using locks on the generator is quite a bottleneck for larger applications. 

Took me a while to realize that exceptions thrown were there because of possible multi-threading issues. I got the same error trace as Threetwosevensixseven, which at first don't seem related to multithreading, I only realized it was due to threading when I wrote some test functions.

**Comment by pulsatrixbv** [comment]
Problem seems to be in the SvgElementFactory class. This class has a static dictionary which is accessed and altered by some static functions in the class. I think that a revision of this class to a non static item will solve the problem.

**Comment by abnersouza** [comment]
Same problem here, using in a Silverlight Application.

I've tried the approach suggested by @kimmccoy still having problems.
If debugging line by line works, but when don't give null reference exception.

**Comment by pulsatrixbv** [comment]
@abnersouza when debugging sometimes multi-threading is disabled (or lesser threads are executed in parallel). Also having the debugger on (or even paused in the code) might prevent raceconditions to occur which caused the issue. The patch I submitted should limit these issues to a level where memory issues start occuring, perhaps try that one?

**Comment by abnersouza** [comment]
@pulsatrixbv Yes so far no issues, gonna do some low testing... and let you know, anyway thanks so much, do you think i can now remove the approach suggested by @kimmccoy? Because I'm still using it. Thans

**Comment by pulsatrixbv** [comment]
You are welcome!

Yep, you can remove the other fix. The old fix disables the entry of concurrent threads by forcing a lock on the render, meaning only one thread can enter.

The patch l submitted should work for multiple threads but I ran into memory issues when processing a lot of images at once (20 concurrent drawings of 5000x5000 pixels). In that case you can limit the maximum number of threads (I use parallel foreach so I put a max number of threads in the execution of the parallel).

Let me know if you need more help for your implementation.


On 27 Aug 2015, at 18:01, abnersouza <notifications@github.com<mailto:notifications@github.com>> wrote:


@pulsatrixbv<https://github.com/pulsatrixbv> Yes so far no issues, gonna do some low testing... and let you know, anyway thanks so much, do you think i can now remove the approach suggested by @kimmccoy<https://github.com/kimmccoy>? Because I'm still using it. Thans

-

**Comment by abnersouza** [comment]
I really appreciate your help.

I'm in a project of Digital Signature, I'm using fabric.js to get the canvas and export to an SVG xml, then i load the svg using XmlDocument, then I use the SvgDocument.Open(sgvXml) to be able to export Draw() and get a Bitmap.

My big problem is that I have to scale the SvgDocument up, before I Draw it, because i need to crop the Bitmap to extract only the signature without the blank area, for that I'm using the approach http://stackoverflow.com/questions/4820212/automatically-trim-a-bitmap-to-minimum-size we were having so many memory issues than we had to implement a mutex to handle that.

I know that i can have from the SVGDocument the bounds of the canvas, that is what i need, but do you know if there is any other way to get only path inside it, without having to crop using that solution mentioned above?

Thanks a lot.

Abner

**Comment by dillanmann** [comment]
I'm still receiving this error, have tried both @kimmccoy and @Threetwosevensixseven solutions but no dice.

The SVG document I am using is here https://pastebin.com/1tmKjQC5 as an example.

**Comment by gvheertum** [comment]
@dillanmann can you tell a bit more about what is going wrong? Are you getting the null-pointer? If so, could you provide a stacktrace or a bit more information (like calling code). I haven't seen the null pointer error in a long time, only issues I sometimes see are out-of-memory exceptions when the canvasses are way to big.

**Comment by dillanmann** [comment]
Apologies for the fairly non-descriptive post!

The version I am using is version 2.2.1.39233, the latest version available via NuGet.

The code that is throwing the error is:

```
        public static void RenderSvgStringToPng(XDocument xmlDoc, string outputFile)
        {
            var innerXml = xmlDoc.ToString();
            using (var xmlStream = new MemoryStream(Encoding.Default.GetBytes(innerXml)))
            {
                xmlStream.Position = 0;
                var doc = SvgDocument.Open<SvgDocument>(xmlStream); // <---- throwing here
                using (var image = doc.Draw())
                    image.Save(outputFile, ImageFormat.Png);
            }
        }
```

Note that it throws the error at SvgDocument.Open regardless of whether I just give it the plain SVG string (from `xmlDoc.ToString()` using `SvgDocument.FromSvg`) or whether I use the memory stream way.

The error I get is:

```
An exception of type 'System.NotSupportedException' occurred in Svg.dll but was not handled in user code

Additional information: Must call AddTerm AddSeparator in that order
```

I noticed some first chance exceptions logged:

```
A first chance exception of type 'System.ArgumentNullException' occurred in Svg.dll
iisexpress.exe Warning: 0 : Attribute 'd' cannot be set - type 'Svg.Pathing.SvgPathSegmentList' cannot convert from string ''.
A first chance exception of type 'System.NotSupportedException' occurred in Svg.dll
```

This is all the detail I could find, let me know anything else you need! :)

**Comment by gvheertum** [comment]
@dillanmann no worries :) Thanks for the details. 

From what I can see it fails on something *in* the SVG rather than crashing on threading issues. Nevertheless, if there is something making the SVG invalid, it should fail with more grace than a nullpointerexceptions. From what I see now (after a quick glance at the problem) is it tries to make a path, but the property is empty (iisexpress.exe Warning: 0 : Attribute 'd' cannot be set - type 'Svg.Pathing.SvgPathSegmentList' cannot convert from string ''.).

I'm not at the machine where I can test this, but I'll try to look into this when I am getting back at my PC and see if I can reproduce this and try to pinpoint what exactly goes wrong.

**Comment by gvheertum** [comment]
Checking the SVG source I noticed:
```html
<path class="plot-now" d="" style="opacity: 0;"></path>
```
Can you try what happens if you remove that element (or provide a d value)? I think this might be the culprit.

However, Chrome and other tools *do* accept this tag, so the SVG generator should not fail on this one. But maybe removing this element might be a temporary fix for your problem :)

**Comment by dillanmann** [comment]
Have tried removing the element but it unfortunately still gives me the same error, except now the first  exception (`iisexpress.exe Warning: 0 : Attribute 'd' cannot be set - type 'Svg.Pathing.SvgPathSegmentList' cannot convert from string ''.`) is no longer present in the output log.

Not sure how much use it'll be but the stack trace for the error (beyond my code) is:

```
   ExCSS.TermList.AddSeparator(GrammarSegment termSepertor) in C:\Dev\vvvv\SVG\Source\External\ExCSS\Model\Values\TermList.cs:62
   ExCSS.Parser.ParseSingleValue(Block token) in C:\Dev\vvvv\SVG\Source\External\ExCSS\Parser.Blocks.cs:389
   ExCSS.Parser.ParseHexValue(Block token) in C:\Dev\vvvv\SVG\Source\External\ExCSS\Parser.Blocks.cs:485
   ExCSS.Parser.ParseTokenBlock(Block token) in C:\Dev\vvvv\SVG\Source\External\ExCSS\Parser.Blocks.cs:103
   ExCSS.Parser.Parse(String css) in C:\Dev\vvvv\SVG\Source\External\ExCSS\Parser.cs:64
   Svg.SvgDocument.Open(XmlReader reader) in C:\Dev\vvvv\SVG\Source\SvgDocument.cs:335
   Svg.SvgDocument.Open(Stream stream, Dictionary`2 entities) in C:\Dev\vvvv\SVG\Source\SvgDocument.cs:237
   Svg.SvgDocument.Open(Stream stream) in C:\Dev\vvvv\SVG\Source\SvgDocument.cs:196
```

**Comment by dillanmann** [comment]
Sorry for the reply spam - From some testing if I remove the entire 'style' element it will work, but obviously the styling is lost. 

I then tried adding the style back in but removing the comments (in case something from the SASS -> CSS generation was causing the issue) but this didn't work.

At least I've managed to tie this down somewhere but don't really understand why the style would be causing the issue. Will keep debugging and let you know, but would appreciate any input you've got too!

**Comment by gvheertum** [comment]
I don't mind the messages (they are not spammy to me), this particular issue caught my interest so the more info the better. Also, the SVG library helped me quite a bit in the past, I would love to give back to the community :)

Probably the error regarding the empty string was thrown, but handled somewhere in the code (first chance exceptions are not always that bad it seems). It could have been an indicator, but for now it doesn't seem to be the cause of the issues you are experiencing. The trace you posted shows the error occurs in the parsing of the CSS, this makes sense, because you said that removing the style stuff gets rid of the error and bringing it back also returns the error.

I noticed in the style-tag that there are > selectors used in the CSS, which are escaped to &amp;gt; tags in the output, Chrome doesn't seem to care, but perhaps these are causing issues for the SVG library. If I find and replace these in the raw SVG you provided (&amp;gt; -> >), the SVG still is rendered correctly in Chrome. So perhaps trying that can solve the issue you are experiencing.

I am wondering if the &gt; notation in styling is something that is allowed by the standard (CSS and/or SVG) or that Chrome is just being friendly/helpful (Chrome sometimes allows/fixes stuff that is out-of-standard, which is helpful when you are having files with this, but really unhelpful if you want to validate behaviour). If it's allowed by the standards, it might be worth fixing it.

**Comment by dillanmann** [comment]
I really appreciate all the effort you're going to here :) 

In regards to the escaped `>` characters, I've actually got some unit tests to try and make it a bit easier to replicate this because I had the same idea. I essentially stripped out all of the css and just placed in potentially erroneous parts but the error was **not** thrown when I added in some css with the escaped characters in it.

**Comment by gvheertum** [comment]
I went a bit deeper and I think I found something. There are some elements in the styling that cause issues. I also added some test cases and found that indeed switching &&amp;gt; and > doesn't matter and indeed removing all styling yields no exception.

Upon closer inspection of the error I found that the `Message: Test method Svg.UnitTests.CssParsingIssue_Issue59.LoadSvg_TestComplexDocumentWithRawChars_YieldsResult threw exception: 
System.NotSupportedException: Must call AddTerm AddSeparator in that order` is caused by the following statement in #footer:  
```css 
border-top: 1px solid #009c46 !important;
```
The problem seems to be the !important element, if I remove this, the above error is no longer shown, however it will fail on other issues, being that pseudo-selectors like :hover and :before are not allowed.

That the !important causes issues is a bug of some sort, since in other elements the !important tag is not causing errors, but in the border-top situation is seems to fail when parsing the value. I'm going to check if I can roll a small fix for this. For the psuedo selectors I'm not sure if we should allow them or how to handle them.

You seem to attach the whole CSS of a website to the image, this makes the CSS to parse quite big, but also introduces elements in the CSS which are not really needed by the SVG (or better said, might even confuse the parser). Regardless of the fix, I would suggest splitting the CSS for the image from the other CSS and server that to the parser if possible.

**Comment by dillanmann** [comment]
Nice that we've found a resolution! :)

I was speaking with the front end dev on my team about this today, the reason the whole site css is includes is because we recently moved to using SASS/LESS and so all of the individual SCSS files are compiled down into a single css file rather than having a site wide and a plot specific one like we had before. Thankfully, I can rectify this fairly easily!

Thanks so much for your help, always nice to see folks willing and able to help out.

I have another issue with text not rendering as expected, but will report this in the appropriate issue / create one if it doesn't exist.

Again, thanks for your help :)

**Comment by gvheertum** [comment]
Good to hear you are happy and seem to have fixed the issue.

There seems to be something really awkward with the parsing of the border properties. I played around with it a bit more and found that
```css
border-top: 1px solid  #009c46  ;
```
failes with the same error `Svg.UnitTests.CssParsingIssue_Issue59.LoadSvg_CssParserCanHandleDoubleWhitespace_YieldsResult threw exception: 
System.NotSupportedException: Must call AddTerm AddSeparator in that order
` while
```css
border-top: 1px solid #009c46;
```
doesn't seem to cause issues.

So the whitespacing seems to cause issues here. My best guess is that the !important is ignored and results in a floating whitespace in the CSS definition, which in his turn borks the parser... I did dive into the parser/lexer (result) of the css and found that indeed some additional whitespace is inserted in the list of tokens for this tag (in case of the extra whitespace and in case of the !important). This behavior does not seem to apply for all tags/definitions (since !important in other parts of your css does not yield any issues.

If I look closely at the result of the lexer in the error case, it seemed to only have parsed the 1px solid and not (yet) the color part. I'm going to look a bit deeper into it, because I am really curious why this happens.

**Comment by mrbean-bremen** [comment]
@gvheertum - I'm trying to understand the state of this old issue.
You have already done a fix and a lot of investigation for this issue - can you summarize the latest state? Is the multi-threading problem still there, and was the problem with `!important` / white spaces independent of this?

**Comment by gvheertum** [comment]
@mrbean-bremen - Wow, this is an oldy. I'll try next week to see if I can trace back what I did here. As far as I know, this was not related to the multi-threading, but more something in the lexer/parser of stylesheets. I remember that it was quite strange behaviour I saw when trying to trace the issue.

The null/fail load for multi-threading was fixed in the multi-threading update and the issue regarding the lexer appeared to not be related to it (it failed even without multi-threading). Somehow the issue Dillan encountered ended up in this thread and was discussed here (possibly because the project was relatively silent back then and this was one of the few active topics, also because it was an issue while loading the file, so I can imagine it might seem related at the moment).

Maybe you can split the lexer/styling issue to a different ticket/issue and close this one. I'll try doing some research next week regarding the lexer issue, but I'm affraid I lost most of the test files/data I used back then, but I'll try to get back to you about this.

By the way, great to see this project getting brought back to life! 😄

**Comment by mrbean-bremen** [comment]
Thank you a lot! I will create a new issue in this case and copy over the relevant parts from this one, so you can comment on the new one - it would be really nice if you can find an example file that shows the problem.

**Comment by mrbean-bremen** [comment]
Closing the original issue, [spin-off](https://github.com/vvvv/SVG/issues/399) has been created.
